fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1263,212 @@ class SamOrder(enum.Enum):
+        return secondaries
+
+    @classmethod
+    def many_sam_from_rec(cls, rec: AlignedSegment) -> Iterator[AlignedSegment]:
+        """Build many SAM secondary alignments from a single SAM record.
+
+        All reconstituted secondary alignments will have the `rh` SAM tag set upon them.
+
+        By default, the query bases and qualities of the secondary alignment will be set to the
+        query bases and qualities of the record that created the secondary alignments. However, if
+        there are hard-clips in the record used to create the secondary alignments, then this
+        function will set the query qualities and bases to the space-saving and/or unknown marker
+        `*`. A future development for this function should correctly pad-out (with No-calls) or clip
+        the query sequence and qualities depending on the hard-clipping found in both ends of the
+        source (often a primary) record and both ends of the destination (secondary) record.
+
+        Args:
+            rec: The SAM record to generate secondary alignments from.
+        """
+        if (
+            rec.is_unmapped
+            or rec.cigarstring is None
+            or rec.query_sequence is None
+            or rec.query_qualities is None
+        ):
+            return
+
+        for hit in cls.many_from_rec(rec):
+            # TODO: When the original record has hard clips we must set the bases and quals to "*".
+            #       It would be smarter to pad/clip the sequence to be compatible with new cigar...
+            if "H" in rec.cigarstring:
+                query_sequence = NO_QUERY_BASES
+                query_qualities = None
+            elif rec.is_forward and not hit.is_forward:
+                query_sequence = reverse_complement(rec.query_sequence)
+                query_qualities = rec.query_qualities[::-1]
+            else:
+                query_sequence = rec.query_sequence
+                query_qualities = rec.query_qualities
+
+            secondary = AlignedSegment(header=rec.header)
+            secondary.query_name = rec.query_name
+            secondary.reference_id = rec.header.get_tid(hit.reference_name)
+            secondary.reference_name = hit.reference_name
+            secondary.reference_start = hit.reference_start
+            secondary.mapping_quality = 0 if hit.mapq is None else hit.mapq
+            secondary.cigarstring = str(hit.cigar)
+            secondary.query_sequence = query_sequence
+            secondary.query_qualities = query_qualities
+            secondary.is_read1 = rec.is_read1
+            secondary.is_read2 = rec.is_read2
+            secondary.is_duplicate = rec.is_duplicate
+            secondary.is_paired = rec.is_paired
+            secondary.is_proper_pair = False
+            secondary.is_qcfail = rec.is_qcfail
+            secondary.is_forward = hit.is_forward
+            secondary.is_secondary = True
+            secondary.is_supplementary = False
```

Comment:
**issue** In the current implementation, `cls.many_from_rec()` rehydrates both secondary alignments from the `XB` tag and supplementary alignments from the `XA` tag into the alignment records processed by this method. Supplementary alignments should be flagged appropriately.

Reply:
Also on this topic - with our usual alignment settings, `bwa` will output alignment records for supplementary alignments. Do we need to add logic to ensure that the rehydration does not create redundant records?

Reply:
See https://github.com/fulcrumgenomics/fgpyo/pull/206#discussion_r1899623157

Supplementaries come from the `SA` tag.

Bwa can be run in a mode where it makes the `XA` tag for secondaries _**or**_ a mode where it makes the `XB` tag for secondaries. I didn't bother implementing logic distinguishing the two since it's highly unlikely a BAM will have both tags on the _**same record**_. But even if you did, I don't see that as a bug, just as an odd use of using both tag formats simultaneously.